### PR TITLE
Split invalidation triggers by source kind; drop trigger-publisher footgun

### PR DESCRIPTION
**Pull request by kibertoad**
## Summary

- Triggers now apply resolved invalidations directly to a `Loader` / `GroupLoader` (`target: …`) instead of republishing through a `NotificationPublisher`. The loader's existing in-memory + async + notification-pair machinery handles everything — no separate trigger publisher, no `serverUuid` juggling.
- The monolithic `SqsInvalidationTrigger` (which used a `sourceType` discriminator and forced `dependencies` to be repeated per source) is replaced by four homogeneous classes:
  - `SnsTopicInvalidationTrigger` / `SqsQueueInvalidationTrigger`
  - `SnsTopicGroupInvalidationTrigger` / `SqsQueueGroupInvalidationTrigger`

  Each takes one `dependencies` block and an array of `sources` — listening to multiple queues/topics is just listing them.
- A source with multiple bindings can route different event types to different resolvers via a `messageTypeField` discriminator. `composeTriggers(...)` bundles multiple triggers for deployments that need to mix source kinds.
- Docs reoriented (incorporating guidance from #398) to recommend Redis pub/sub as the default and to lead with the hybrid pattern — Redis fanout in the Loader plus an SQS trigger on a **shared** upstream queue. The new target-based trigger API makes that shape trivially work without any queue-lifecycle plumbing.

## Why

The previous design republished trigger output through SNS to reach every instance — including the originating one. That required the trigger's publisher to use a `serverUuid` distinct from the local pair, or the local consumer would skip its messages as self-emitted and silently leave the in-memory cache stale. The loader already knows how to apply invalidations correctly and fan them out via its own publisher, so the round-trip and the footgun both disappear when triggers target the loader instead.

Splitting by source kind drops the `sourceType` discriminator and lets each trigger class declare exactly the `@message-queue-toolkit` dependency shape it needs (SQS-only vs. SNS-SQS), so `dependencies` lives at the trigger level instead of being copied per source.

## Breaking changes

- `SqsInvalidationTrigger` and `SqsGroupInvalidationTrigger` are removed. Migrate by:
  - Picking `SnsTopicInvalidationTrigger` or `SqsQueueInvalidationTrigger` (group variants where relevant) based on which `sourceType` you had.
  - Replacing the top-level source fields with `sources: [{ creationConfig | locatorConfig, bindings: [{ messageSchema, resolver }] }]`.
  - Replacing `publisher: triggerPublisher` with `target: yourLoader`. Delete the trigger publisher entirely — no separate publisher and no `randomUUID()` server UUID required.
- The `set` invalidation action kind is dropped from `InvalidationAction`. Triggers should emit `delete` / `deleteMany` and let the loader refetch on demand. (`set` had no clean equivalent on the loader's public API and was a poor fit for an external trigger.)

## Test plan

- [x] `tsc --noEmit` clean for both the root and `packages/sqs`.
- [x] All 27 SQS package tests pass (`vitest run` in `packages/sqs`), including new coverage for: multiple SNS-topic sources, multiple SQS-queue sources, multi-binding routing by `messageTypeField`, and `composeTriggers` driving an SNS + SQS pair as one unit.
- [x] All 259 root tests pass (`vitest run` at the root).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Documentation**
  * Revamped notification docs with clearer Redis vs AWS SNS/SQS guidance, three deployment profiles (Pure Redis / Pure SNS-SQS / Hybrid), a recommended Redis-fanout + SQS pattern, and an explicit subscribe→validate→resolve→apply trigger flow with updated examples.

* **New Features**
  * Added dedicated SNS and SQS trigger types, flat/group binding helpers, and ability to compose/run multiple triggers concurrently.

* **Behavior Change**
  * Removed the "set" invalidation action; only delete/deleteMany/clear remain.

* **Infrastructure**
  * Raised Node engine requirement and updated CI/test matrix; tests adjusted for new trigger lifecycle and composition.

[Review Change Stack](https://app.coderabbit.ai/change-stack/kibertoad/layered-loader/pull/399?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: packages/sqs/lib/triggers/AbstractSqsTrigger.ts [hunk 1]
```diff
@@ -124,66 +89,31 @@ export abstract class AbstractSqsTrigger<TMessage extends object, TContext>
...
+export function composeTriggers(
+  ...triggers: readonly InvalidationTrigger[]
+): InvalidationTrigger {
+  return {
+    async start() {
+      await Promise.all(triggers.map((t) => t.start()))
+    },
+    async stop() {
+      await Promise.all(triggers.map((t) => t.stop()))
+    },
```
[the remark this answers is not in the document]
Comment (kibertoad) [reply packages/sqs/lib/triggers/AbstractSqsTrigger.ts:145]: Applied in 1c91be8 — switched to `Promise.allSettled` in both `AbstractSqsTrigger.stop()` and `composeTriggers().stop()` so every underlying consumer/trigger gets a chance to clean up even when one rejects.

File: packages/sqs/test/SqsInvalidationTrigger.spec.ts [hunk 2]
```diff
@@ -33,506 +33,1003 @@ describe('SqsInvalidationTrigger', () => {
...
-      const upstreamTopicArn = createTopic.TopicArn!
 
-      const triggerPublisher = new SqsNotificationPublisher<string>({
-        serverUuid: randomUUID(),
-        dependencies: buildPublisherDeps(clients),
-        locatorConfig: { topicName: invalidationTopic },
-      })
+      await waitFor(() => resolverCalls > 0)
+      await new Promise((r) => setTimeout(r, 200))
+      expect(loader.getInMemoryOnly('user-9')).toBe('value')
```
[the remark this answers is not in the document]
Comment (kibertoad) [reply packages/sqs/test/SqsInvalidationTrigger.spec.ts:312]: Declining for now. The test is asserting a negative — that a `null`-returning resolver does **not** invalidate. `waitFor(() => resolverCalls > 0)` already proves the message reached `runFlatPipeline`, so any invalidation would happen synchronously inside it before the resolver returns `null`. The 200 ms is a conservative buffer for downstream cache writes. Polling the cache value over a 300–1000 ms window would inflate this single test by 5× without catching anything the synchronous path can miss. Happy to revisit if we see real flakes.

**Comment by kibertoad** [comment]
Addressing CodeRabbit feedback in 1c91be8 (plus CI pin in 34fa7d5):

**Applied**

- *`AbstractSqsTrigger.ts` — `Promise.allSettled` in `stop()`*: switched both `AbstractSqsTrigger.stop()` (line 98) and `composeTriggers().stop()` (line 116) to `Promise.allSettled` so every consumer/trigger gets a shutdown attempt even if one rejects.
- *`packages/sqs/README.md` outside-diff staleness*: the lifecycle example now uses `SnsTopicInvalidationTrigger` instead of the removed `SqsInvalidationTrigger`, and the `SqsNotificationPublisher` API-reference row no longer talks about "trigger publishers" (that concept went away with the refactor).

**Declined**

- *`SqsInvalidationTrigger.spec.ts:310-312` — fixed 200 ms sleep*: replying inline. Test asserts a *negative* (null resolver ⇒ no invalidation); `waitFor(resolverCalls > 0)` already proves the message ran through `runFlatPipeline` synchronously. Polling for 300–1000 ms would 5× the test runtime without catching anything the synchronous path can miss.
- *`bindingHelpers.ts:86-89` — edge case when `messageTypeField` is set on a single binding*: skipping per CodeRabbit's own "💤 Low value" tag. With a single binding the `MessageHandlerConfigBuilder` accepts the un-discriminated handler and the runtime `messageTypePath` resolves dynamically; the cast yields an unused `undefined` `messageType` option. Will tighten validation if a real user hits this.
- *`bindingHelpers.ts:100-103, 148-152` — `err as Error` cast*: skipping, also flagged as "💤 Low value". Throwing non-`Error` values would be exotic for this code path; `errorHandler` consumers already have to handle structured logging on whatever they receive.

CI pin: `pnpm@11.1.2` (introduced in #396) requires Node ≥22.13 and was breaking the Node 20.x leg of the CI matrix. Rolled `packageManager` back to `pnpm@10.33.0` so all four matrix entries install.

File: packages/sqs/lib/triggers/AbstractSqsTrigger.ts [hunk 3]
```diff
@@ -124,66 +89,31 @@ export abstract class AbstractSqsTrigger<TMessage extends object, TContext>
...
-    if (source.creationConfig?.queue?.QueueName) return source.creationConfig.queue.QueueName
-    if (source.locatorConfig?.queueName) return source.locatorConfig.queueName
-    if (source.locatorConfig?.queueUrl) return source.locatorConfig.queueUrl
-    return 'sqs-invalidation-trigger'
+export function composeTriggers(
+  ...triggers: readonly InvalidationTrigger[]
+): InvalidationTrigger {
+  return {
+    async start() {
+      await Promise.all(triggers.map((t) => t.start()))
```
Comment (CarlosGamero) [inline packages/sqs/lib/triggers/AbstractSqsTrigger.ts]: Not sure, but could we have a leak here? I mean, trigger A starts successfully and trigger B throws, the composite rejects but A is left running.

Should we follow the same patten as the above `start` method with a `try/catch`?

File: packages/sqs/lib/triggers/AbstractSqsTrigger.ts [hunk 4]
```diff
@@ -124,66 +89,31 @@ export abstract class AbstractSqsTrigger<TMessage extends object, TContext>
...
+    },
+    async stop() {
+      await Promise.allSettled(triggers.map((t) => t.stop()))
```
Comment (CarlosGamero) [inline packages/sqs/lib/triggers/AbstractSqsTrigger.ts]: Should we ignore errors here?

File: packages/sqs/lib/triggers/AbstractSqsTrigger.ts (hunk 3 above)
Comment (kibertoad) [reply packages/sqs/lib/triggers/AbstractSqsTrigger.ts]: Good catch — fixed in 8b2c9d0. `composeTriggers().start()` now collects rejections from `Promise.allSettled` and, if any failed, stops the triggers that *did* start before rethrowing. Multiple failures get wrapped in `AggregateError`. Kept the parallel fan-out semantics that the README documents.

File: packages/sqs/lib/triggers/AbstractSqsTrigger.ts (hunk 4 above)
Comment (kibertoad) [reply packages/sqs/lib/triggers/AbstractSqsTrigger.ts]: Mostly yes, with a nuance — 8b2c9d0 splits the two callsites:

- `AbstractSqsTrigger.stop()` (closing the underlying mqt consumers): still swallows. Shutdown is best-effort, the caller has no recovery, and the underlying SQS/SNS connection may already be torn down — logging here would just be noise. Added a comment documenting that.
- `composeTriggers().stop()` (operating on whole triggers): now logs rejections via `console.error` so a silent failure during shutdown isn't invisible, but still doesn't throw — same rationale (best-effort cleanup, no useful caller recovery).

File: packages/sqs/lib/triggers/bindingHelpers.ts [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+  messageTypeField: string | undefined,
+  target: InvalidationTarget,
+  channel: string,
+  errorHandler: TriggerErrorHandler | undefined,
+): BuiltBindings<InvalidationTarget> {
+  validateBindings(bindings, messageTypeField, channel)
+
+  const messageTypeResolver: MessageTypeResolver = messageTypeField
+    ? { messageTypePath: messageTypeField }
+    : { literal: bindings[0]!.messageType ?? DEFAULT_FLAT_MESSAGE_TYPE }
```
Comment (CarlosGamero) [inline packages/sqs/lib/triggers/bindingHelpers.ts]: Maybe using `?` instead of `!` here
```suggestion
    : { literal: bindings[0]?.messageType ?? DEFAULT_FLAT_MESSAGE_TYPE }
```

File: packages/sqs/lib/triggers/bindingHelpers.ts [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+  const context: BindingHandlerContext<InvalidationTarget> = { target, channel, errorHandler }
+
+  const builder = new MessageHandlerConfigBuilder<object, BindingHandlerContext<InvalidationTarget>>()
+  for (const binding of bindings) {
+    const messageType =
+      binding.messageType ?? (messageTypeResolver as { literal: string }).literal
```
Comment (CarlosGamero) [inline packages/sqs/lib/triggers/bindingHelpers.ts]: Mmm, is this casting actually safe?, I mean `messageTypeResolver` can contain `messageTypePath` instead of `literal` in some cases right?

File: packages/sqs/lib/triggers/types.ts [hunk 7]
```diff
@@ -32,16 +32,37 @@ export type InvalidationResolver<TMessage, TAction> = (
...
 ) => ResolverOutput<TAction> | Promise<ResolverOutput<TAction>>
 
+/**
+ * Minimal surface a flat trigger needs from a `Loader` (or any other cache
+ * facade) to apply resolved {@link InvalidationAction}s locally. `Loader`
+ * satisfies this structurally.
+ */
+export interface InvalidationTarget {
+  invalidateCacheFor(key: string): Promise<void>
+  invalidateCacheForMany(keys: string[]): Promise<void>
```
Comment (CarlosGamero) [inline packages/sqs/lib/triggers/types.ts]: Maybe we can use `readonly` here? 

```suggestion
  invalidateCacheForMany(keys: readonly string[]): Promise<void>
```
So `dispatch.ts:38` doesn't need the defensive [...action.keys] spread

File: packages/sqs/lib/triggers/AbstractSqsTrigger.ts [hunk 8]
```diff
@@ -87,35 +52,35 @@ class SnsTopicTriggerConsumer<TMessage extends object, TContext> extends Abstrac
...
+  protected abstract createConsumers(): readonly InternalConsumerHandle[]
 
   async start(): Promise<void> {
     if (this.startPromise) return this.startPromise
-    if (this.internalConsumer) return
+    if (this.internalConsumers.length > 0) return
     this.startPromise = (async () => {
-      const consumer = this.createConsumer()
+      const consumers = this.createConsumers()
+      if (consumers.length === 0) return
```
Comment (CarlosGamero) [inline packages/sqs/lib/triggers/AbstractSqsTrigger.ts]: Could `this.createConsumers` throw? should we include withint the try/catch to ensure finally is executed and the promise is cleaned

**Review by CarlosGamero — APPROVED** [review]

File: packages/sqs/lib/triggers/bindingHelpers.ts (hunk 5 above)
Comment (kibertoad) [reply packages/sqs/lib/triggers/bindingHelpers.ts]: Applied in 4e14039 — switched to optional chaining in both `buildFlatBindings` and `buildGroupBindings`.

File: packages/sqs/lib/triggers/bindingHelpers.ts (hunk 6 above)
Comment (kibertoad) [reply packages/sqs/lib/triggers/bindingHelpers.ts]: Good question — the cast was technically safe (when `messageTypeField` is set, `validateBindings` requires every binding to carry its own `messageType`, so `binding.messageType ?? …` always short-circuits before hitting the fallback), but that invariant wasn't obvious from the call site. Refactored in 4e14039 to extract a single `fallbackMessageType` constant that's reused for both the resolver and the per-handler `messageType`, removing the cast entirely.

File: packages/sqs/lib/triggers/types.ts (hunk 7 above)
Comment (kibertoad) [reply packages/sqs/lib/triggers/types.ts]: Done in 4e14039 — `InvalidationTarget.invalidateCacheForMany` now takes `readonly string[]` and `applyFlatAction` forwards `action.keys` directly. `Loader.invalidateCacheForMany` already only reads the array, so the structural compatibility holds without changes on the root package side.

File: packages/sqs/lib/triggers/AbstractSqsTrigger.ts (hunk 8 above)
Comment (kibertoad) [reply packages/sqs/lib/triggers/AbstractSqsTrigger.ts]: Yes — `createConsumers()` calls into `buildFlatBindings`/`buildGroupBindings`, which call `validateBindings`, so it can synchronously throw on misconfigured sources. Currently if it does, `startPromise` stays set to a rejected promise and the trigger is wedged for the rest of its life.

Fixed in 4e14039 by moving `createConsumers()` inside an outer `try/finally` (the existing inner `try/catch` still handles the close-on-error rollback for already-init'd consumers). `startPromise` is now cleared in both paths.
